osu.Game/Updater/AssociationManager.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            Logger.Log("Setting up file associations!");
+            string programPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location) + "\\osu!.exe";
+
+            foreach (string extension in associated_extensions)
+            {
+                CreateFileAssociation(extension, programPath);
+            }
+
+            SHChangeNotify(0x8000000, 0x1000, IntPtr.Zero, IntPtr.Zero); // Notify Explorer that the file associations have changed
+        }
+
+        /// <summary>
+        /// Create an association for a specific extension
+        /// </summary>
+        /// <param name="extension"></param>
+        /// <param name="programPath"></param>
+        /// <exception cref="InvalidOperationException"></exception>
+        public void CreateFileAssociation(string extension, string programPath)
+        {
+            Logger.Log("Creating file association for " + extension);
+
+            // Check if the extension has a file explorer override, if so, delete the override
+            RegistryKey? fileExplorerKey = Registry.CurrentUser.OpenSubKey("Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\FileExts\\" + extension, true);
+
+            if (fileExplorerKey != null)
+            {
+                fileExplorerKey.DeleteSubKey("UserChoice", false);
+                fileExplorerKey.Close();
+                Logger.Log("Deleted file explorer override for " + extension);
+            }
+            else
+            {
+                Logger.Log("File explorer override for " + extension + " does not exist!");
+            }
+
+            RegistryKey key = Registry.CurrentUser.CreateSubKey($"Software\\Classes\\{extension}");
+            key.CreateSubKey("shell\\\\open\\\\command").SetValue("", $"\"{programPath}\" \"%1\"");
+            key.Close();
+            Logger.Log("Created file association for " + extension);
+        }
+
+        public void RemoveFileAssociation(string extension)
+        {
+            Logger.Log("Removing file association for " + extension);
+            RegistryKey? key = Registry.CurrentUser.OpenSubKey($"Software\\Classes\\{extension}");
+
+            if (key != null)
+            {
+                Registry.CurrentUser.DeleteSubKeyTree($"Software\\Classes\\{extension}");
+                Logger.Log("Removed file association for " + extension);
+            }
+            else
+            {
+                Logger.Log("File association for " + extension + " does not exist!");
+            }
+        }
+
+        /// <summary>
+        /// Removes all of the file associations
+        /// </summary>
+        public void RemoveFileAssociations()
+        {
+            foreach (string extension in associated_extensions)
+            {
+                RemoveFileAssociation(extension);
+            }
+
+            SHChangeNotify(0x8000000, 0x1000, IntPtr.Zero, IntPtr.Zero); // Notify Explorer that the file associations have changed
```

Comment:
I think it might be wise to move this to `RemoveFileAssociation()`. I added this here thinking it would be better to just refresh explorer after all operations are finished, but this would technically go against the rules of reuse since calling `RemoveFileAssociation()` directly (i.e. for just one extension) would remove the association, but not refresh explorer.